librosa/core/pitch.py

```diff
@@ -335,3 +338,281 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
                                   + dskew[idx[:, 0], idx[:, 1]])
 
     return pitches, mags
+
+
+def yin(y, sr=22050, win_length=1024, hop_length=None, fmin=55, fmax=900, peak_threshold=0.1):
```

Comment:
Will do. Would you be happy with setting `win_length = 2048` and then dividing by 2 for the autocorrelation calculation? Also, do you have any preferences for a default value for `fmax`?

Reply:
Just a note from the pYIN paper:
> The pitch tracking step consists of choosing at most one pitch candidate at every frame. We divide the pitch space into 480 bins ranging over four octaves from 55Hz (A1) to just under 880Hz (A5) in steps of 10 cents (0.1 semitones).

Reply:
> Will do. Would you be happy with setting `win_length = 2048` and then dividing by 2 for the autocorrelation calculation?

I gave this a think this morning, and the following seems like the most internally consistent way to parametrize this:

1. Add a `frame_length` parameter (def=2048) that controls the literal frame length
2. Keep `win_length=1024` as the parameter which controls how long the correlation template for each frame is.  `win_length` cannot exceed `frame_length`.  This is basically how it's set up in `stft`, and i think the same reasoning applies here (windows fit inside frames, after all!).

This way, you don't have to worry about any `2*win_length` business.



> Also, do you have any preferences for a default value for `fmax`?

The current settings (inherited from yin) are 4 octaves, which seems pretty narrow to me.  CQT defaults to 7 octaves up from `fmin`, so `C8=4186 Hz`.  If yin is stable up at that range, I'd say to use that, but I'm okay with bringing it down for stability purposes.

Reply:
I did a quick test using a chirp from `C1` to `C8` and it seems like a large pitch range causes octave errors in `yin` at both high and low frequencies. `pyin` on the other hand appears to be relatively stable. However, it is worth noting that increasing the range will add many more states to the HMM and could slow things down considerably. So it might be worth defaulting to a relatively small range and then pointing out in the docs that `fmin` and `fmax` should be adjusted depending on the use-case?
![fmax](https://user-images.githubusercontent.com/24506100/84133613-a83a1480-aa47-11ea-9b7e-398c25fc7fa9.png)

Reply:
Great -- a couple of points here:

1. I'm not too worried about efficiency here, how much slower could it be?  I'd be more worried about memory consumption, but my hunch is that this isn't any worse than a default-sized stft, so we're probably okay.
2. In this example you've posted, is it using centered frames or left-aligned?  Just trying to get a sense of whether the glitches at the signal start are actually due to boundary effects from padding, or if it's a real error.   Padding out the chirp with leading silence might help tease that apart.
3. The pyin curve looks great, so I think we'd be okay to make these the defaults.  In general, I think we should nudge folks away from yin and toward pyin, and it makes more sense to prioritize getting the parameters right for pyin than to tune them for yin and have inconsistent settings.

Reply:
1. I haven't checked memory consumption at all but it hasn't been an issue at all so far. As for efficiency, the Viterbi decoding is `O(T x  S^2)` where T is the sequence length and S is the number of states. Using a range of `C1` to `C8` as opposed to `A1` to `A5` results in roughly twice as many states so I guess you'd get a slow down of about 4x. I think that's probably fine. If a user wants to get better performance then they can adjust things.
2. That plot is left-aligned so I think the errors at the beginning are actual octave errors as opposed to padding artifacts.
3. Yeah, pyin is definitely better in general. I guess you don't want different defaults for pyin vs yin?

Reply:
>     1. Using a range of `C1` to `C8` as opposed to `A1` to `A5` results in roughly twice as many states so I guess you'd get a slow down of about 4x. I think that's probably fine. If a user wants to get better performance then they can adjust things.

Indeed.  The hard work here is in the viterbi algorithm, which is already pretty well accelerate, and I don't expect any problems there.



>     2\. That plot is left-aligned so I think the errors at the beginning are actual octave errors as opposed to padding artifacts.

Ok, thanks for verifying!



>     3\. Yeah, pyin is definitely better in general. I guess you don't want different defaults for pyin vs yin?

Yeah, the defaults should be the same.  API consistency is important here, both for keeping things simple on the user-side (principle of least surprise), and also for interacting with the `presets` library when users want to change defaults.